thrall/app/controllers/ReaperController.scala

```diff
@@ -0,0 +1,195 @@
+    scheduler.scheduleAtFixedRate(
+      initialDelay = 0 seconds,
+      interval = INTERVAL,
+    ){ () =>
+      if(store.client.doesObjectExist(reaperBucket, CONTROL_FILE_NAME)) {
+        logger.info("Reaper is paused")
+      } else {
+        es.countImagesIngestedInLast(7 days)(DateTime.now(DateTimeZone.UTC)).flatMap { imagesIngestedInLast7Days =>
+
+          val imagesIngestedPer15Mins = imagesIngestedInLast7Days / REAPS_PER_WEEK
+
+          val countOfImagesToReap = Math.min(imagesIngestedPer15Mins, 1000).toInt
+
+          if (countOfImagesToReap == 1000) {
+            logger.warn(s"Reaper is reaping at maximum rate of 1000 images per $INTERVAL. If this persists, the INTERVAL will need to become more frequent.")
+          }
+
+          val deletedBy = "reaper"
+
+          Future.sequence(Seq(
+            doBatchSoftReap(countOfImagesToReap, deletedBy),
+            doBatchHardReap(countOfImagesToReap, deletedBy)
+          ))
+        }
+      }
+    }
+  }
+
+  private def batchDeleteWrapper(count: Int)(func: (Int, String) => Future[JsValue]) = auth.async { request =>
+    if (!authorisation.hasPermissionTo(DeleteImage)(request.user)) {
+      Future.successful(Forbidden)
+    }
+    else if (count > 1000) {
+      Future.successful(BadRequest("Too many IDs. Maximum 1000."))
+    }
+    else {
+      func(
+        count,
+        request.user.accessor.identity
+      ).map(Ok(_))
+    }
+  }
+
+  private def persistedBatchDeleteOperation(deleteType: String)(doBatchDelete: => Future[JsValue]) = config.maybeReaperBucket match {
+    case None => Future.failed(new Exception("Reaper bucket not configured"))
+    case Some(reaperBucket) => doBatchDelete.map { json =>
+      val now = DateTime.now(DateTimeZone.UTC)
+      val key = s"$deleteType/${now.toString("YYYY-MM-dd")}/$deleteType-${now.toString()}.json"
+      store.client.putObject(reaperBucket, key, json.toString())
+      json
+    }
+  }
+
+  def doBatchSoftReap(count: Int): Action[AnyContent] = batchDeleteWrapper(count)(doBatchSoftReap)
+
+  def doBatchSoftReap(count: Int, deletedBy: String): Future[JsValue] = persistedBatchDeleteOperation("soft"){
+
+    logger.info(s"Soft deleting next $count images...")
+
+    val deleteTime = DateTime.now(DateTimeZone.UTC)
+
+    (for {
+      idsSoftDeletedInES: Set[String] <- es.softDeleteNextBatchOfImages(isReapable, count, SoftDeletedMetadata(deleteTime, deletedBy))
+      idsNotProcessedInDynamo <- softDeletedMetadataTable.setStatuses(idsSoftDeletedInES.map(
+        ImageStatusRecord(
+          _,
+          deletedBy,
+          deleteTime = deleteTime.toString,
+          isDeleted = true
+        )
+      ))
+    } yield idsSoftDeletedInES.map { id =>
+      val detail = Map(
+        "ES" -> true,
+        "dynamo.softDelete.metadata" -> !idsNotProcessedInDynamo.contains(id)
+      )
+      logger.info(s"Soft deleted image $id : $detail")
+      metrics.softReaped.increment().run
```

Comment:
it looks like you can pass a number to increment, so you should be able to pass the number reaped in a batch in a single datapoint, which should be cheaper and faster (PutMetricData calls are $0.01 per 1000, so (for both soft & hard) $20 per million... not really expensive but not cheap either)

Reply:
nice - thanks